bin/seedUserStreaks.ts

```diff
@@ -0,0 +1,179 @@
+import fastq from 'fastq';
+import { differenceInDays } from 'date-fns';
+
+import '../src/config';
+import createOrGetConnection from '../src/db';
+import { User, UserStreak, View } from '../src/entity';
+import { DataSource, DeepPartial, Equal, IsNull, MoreThan, Not } from 'typeorm';
+import { DayOfWeek as Day } from '../src/types';
+
+const QUEUE_CONCURRENCY = 1;
+const BATCH_SIZE = 1000;
+
+interface UserPointer {
+  id?: string;
+  createdAt?: string;
+}
+
+const addNextBatch = async (
+  con: DataSource,
+  insertQueue: fastq.queueAsPromised<Partial<User>[], void>,
+  userPointer?: UserPointer,
+  limit: number = BATCH_SIZE,
+): Promise<UserPointer> => {
+  const qb = con
+    .getRepository(User)
+    .createQueryBuilder('u')
+    // fetching createdAt as a string, because Date object in node is milisecond precision, but postgres is microsecond
+    .select(['id', 'timezone', '"createdAt"::text'])
+    .where(
+      userPointer?.createdAt
+        ? {
+            createdAt: MoreThan(userPointer?.createdAt),
+          }
+        : 'TRUE',
+    )
+    .orWhere(
+      userPointer?.id
+        ? {
+            createdAt: Equal(userPointer.createdAt),
+            id: MoreThan(userPointer.id),
+          }
+        : 'TRUE',
+    )
+    .orderBy('u."createdAt"', 'ASC')
+    .addOrderBy('u."id"', 'ASC')
+    .take(limit);
+
+  const users = await qb.execute();
+
+  // nothing to do anymore, all the users are processed
+  if (users.length === 0) {
+    console.log('no users to process');
+    return null;
+  }
+
+  console.log(
+    'Queuing batch of users to process',
+    users.map((u) => u.id),
+  );
+  insertQueue.push(users);
+
+  const lastUser = users[users.length - 1];
+  return {
+    id: lastUser.id,
+    createdAt: lastUser.createdAt,
+  };
+};
+
+interface ViewWithDateTz extends View {
+  date: Date;
+  maxTimestamp: Date;
+}
+
+const FREEZE_DAYS_IN_A_WEEK = 2;
+const MISSED_LIMIT = 1;
+
+const computeReadingStreaksData = async (
+  con: DataSource,
+  user: Partial<User>,
+): Promise<DeepPartial<UserStreak>> => {
+  const timeZone = user.timezone ?? 'UTC';
+
+  const views: ViewWithDateTz[] = await con
+    .getRepository(View)
+    .createQueryBuilder()
+    .select([])
+    .addSelect('MAX(timestamp)', 'maxTimestamp')
+    .addSelect('DATE(timestamp::timestamptz AT TIME ZONE :timezone)', 'date')
+    .where({ userId: user.id, timestamp: Not(IsNull()) })
+    .innerJoin(User, 'u', `"userId" = u.id`)
```

Comment:
Do we need this join? I'm not sure

Reply:
Not really. Originally it was there because of the date/timezone conversion. But since we supply the timezone with the user as a parameter now, no need for a join anymore.